[release-0.45.x] Release RC artifacts directly to Sonatype

im-konge opened the pull request:
### Type of change

- Enhancement

### Description

As there is no staging repository in new Sonatype portal, we decided to release RCs directly into the repository and not using any other staging repository.
This PR changes the AZP to adapt to this new approach.

In the build pipeline, only in case that we will be on `main` branch the SNAPSHOTS will be pushed.
In case of the release pipeline, we will change the version in poms and everything to the specified `RELEASE_VERSION` and build the code again (from the same commit), which will be in the next step released to Maven repository.

### Checklist

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

scholzj commented on .azure/templates/jobs/build/release_artifacts.yaml:
```diff
@@ -11,13 +11,21 @@ jobs:
         env:
           RELEASE_VERSION: '${{ parameters.releaseVersion }}'
 
+      # Build Java code with changed release version
```
Where exactly do you change the version?

im-konge replied:
it's part of the `make release` goal, where the version in the Maven pom.xmls is set to the one specified by `RELEASE_VERSION`.
My thinking is that it will be still there in the next step.

scholzj commented on .azure/templates/jobs/build/release_artifacts.yaml:
```diff
@@ -11,13 +11,21 @@ jobs:
+        displayName: "Build Java code"
+        env:
+          MVN_ARGS: "-DskipTests -e -V -B"
+          CURL_ARGS: "-s"
```
What does it need `curl` for?

im-konge replied:
Copy paste from different pipeline, I will remove it, thanks.

scholzj reviewed (approved):
I guess this migth work ...

scholzj commented on .azure/templates/jobs/build/deploy_strimzi_java.yaml:
```diff
@@ -23,6 +23,14 @@ jobs:
         parameters:
           JDK_VERSION: $(jdk_version)
 
+      # Change release version in pom file
+      - bash: "make release_maven"
+        displayName: "Change release version in pom files"
+        env:
+          RELEASE_VERSION: '${{ parameters.releaseVersion }}'
+        # We want to change it only in case of release (RC or GA), not for SNAPSHOTs
+        condition: eq(variables['build.sourceBranch'], 'refs/heads/release-')
```
I'm still a bit confused why is this needed. As you said before, `make release` does this, or? Also, You probably do not want to run this on `main` branch. So if it is needed, it should be decoupled from this job?

im-konge replied:
This is running in different job, where the Strimzi repository for the branch is newly checked-out. That means even if you run it somewhere else, you will have to store the whole repository (and the changes) and then download them here and used them.
I hit this when I was testing the changes. Even when the `make release` was executed in the previous step, the changes were not present here.

> Also, You probably do not want to run this on main branch. So if it is needed, it should be decoupled from this job?
Yes, I have the condition there which will run this only in case of the release branch.

scholzj replied:
I think this probably deserves more thinking and consideration.

im-konge replied:
WDYM? You don't like the idea of having the version set there? Intead of doing this, I can add just `make pushnexus` to the release artifacts job. Or you want to have it completely separate?

im-konge replied:
Yes, I can do that, in that case I wouldn't need to change the deploy to maven job, but I will just add the `make pushtonexus` into the release-artifacts job.

im-konge commented:
@ppatierno @scholzj I removed the `make release_maven` from the deploy_strimzi job and added the `make pushtonexus` to release_artifacts job. In that job we are doing the `make release`, so the new version should be properly propagated where we need it. Please let me know if it makes sense to you. Thanks

scholzj commented on .azure/templates/jobs/build/release_artifacts.yaml:
```diff
@@ -1,6 +1,12 @@
 jobs:
   - job: 'release_artifacts'
     displayName: 'Prepare release artifacts'
+    # Strategy for the job
+    strategy:
+      matrix:
+        'deploy_strimzi_java-17':
+          jdk_version: '17'
+          main_build: 'true'
```
I do not think you want to have any matrix here. This will never be needed to run on multiple Java versions as you will release only from one of them.

im-konge replied:
Yeah good point, thanks. I removed this with the `JDK_VERSION` parameter, so the default is used from the `install_java.yaml`.
I wanted to keep it consistent with other things, but I totally agree that it doesn't make sense. Thanks

im-konge commented:
@scholzj @ppatierno JFYI I had to do another small change to `install_java.yaml` to fix issue with the default value used in the configure Java step. When the `parameters` section wasn't specified in the template step, I was getting:

```
##[error]Java $(JDK_VERSION) is not preinstalled on this agent
##[error]Java $(JDK_VERSION) is not preinstalled on this agent
```

The issue was because we used environment variable and not parameter.

I tried this change with the default (which works now) and also tried to configure it as we do in other jobs/pipelines: https://github.com/strimzi/strimzi-kafka-operator/pull/11508/commits/314ab2f36d0f7b89af0f126938022eef44daa84c .
Everything seems to be right: https://dev.azure.com/cncf/strimzi/_build/results?buildId=185112&view=results

I'm tagging you because you already approved this PR. Sorry for the issues.

ppatierno commented:
We are holding on merging this PR. We want to proceed with a proper release having the bridge and oauth fixed as well (there are some CVEs).

scholzj commented:
> We are holding on merging this PR. We want to proceed with a proper release having the bridge and oauth fixed as well (there are some CVEs).

Out of curiosity ... why does merging this PR need to wait for other releases? Merging it does not release anything ... especially since yesterday you insisted on merging incomplete PRs in other repo :-D.
